py/malloc.c

```diff
@@ -287,16 +303,36 @@ void m_tracked_free(void *ptr_in) {
     } else {
         MP_STATE_VM(m_tracked_head) = node->next;
     }
-    m_free(node
-        #if MICROPY_MALLOC_USES_ALLOCATED_SIZE
-        #if MICROPY_TRACKED_ALLOC_STORE_SIZE
-        , node->size
-        #else
-        , gc_nbytes(node)
-        #endif
-        #endif
-        );
+    #if MICROPY_ENABLE_FINALISER
+    if (MP_STATE_THREAD(gc_lock_depth) == 0) {
+        m_tracked_inner_free_node(node);
+    } else {
+        // If GC is already locked (i.e. we're being called from a finaliser)
+        // then can't free immediately. Add the node to a list to be explicitly
+        // freed after GC is finished, so it doesn't have to wait until the next GC
+        // pass.
+        m_tracked_node_insert(&MP_STATE_VM(m_tracked_pending_free_head), node);
+    }
+    #else
+    m_tracked_inner_free_node(node);
+    #endif // MICROPY_ENABLE_FINALISER
+}
+
+#if MICROPY_ENABLE_FINALISER
+void m_tracked_free_pending(void) {
+    m_tracked_node_t *head;
+
+    if (MP_STATE_THREAD(gc_lock_depth) == 0) {
+        while ((head = MP_STATE_VM(m_tracked_pending_free_head))) {
+            MP_STATE_VM(m_tracked_pending_free_head) = head->next;
+            // During gc_sweep_all() these nodes are already free, so check first
+            if (gc_nbytes(head)) {
```

Comment:
I see why you needed to do this.  It feels a bit fragile but I think would be robust in practice.

You actually hit an issue which cropped up before in a different way, and may crop up again: `gc_sweep_all()` is too aggressive in the way it frees.  It should probably be changed to:
- first pass: run finalisers of all alive objects on the GC heap
- second pass: free all remaining memory on the GC heap

Bu, that's for another time.

Reply:
That seems like a good idea, safer all around for finalisers.